tracing-subscriber/src/registry/sharded.rs

```diff
@@ -604,8 +604,6 @@ mod tests {
         state.assert_removed("span1");
         state.assert_removed("span2");
 
-        // Ensure the registry itself outlives the span.
-        drop(dispatch);
```

Comment:
Despite the good comment above I am not sure I fully understand why this was here so I'm not 100% sure removing it is ok. (Same in the test below)

Reply:
I think this is actually pretty important. This is ensuring that the `Registry` (which the `Dispatch` is a reference-counted pointer to) is not dropped until after we make assertions that the spans themselves were dropped. The `Registry` stores all the data associated with the spans, so if it is dropped, it will also drop all the span data. Then, our assertions that _specific_ spans were dropped when they closed will _always_ pass, even if the `Registry` behavior for dropping spans on close was incorrect.

The change to pass the `Dispatch` by value rather than by reference breaks this, since it is now dropped at the end of the `with_default` closure.

I think what we need to do instead is call `Dispatch::clone`, and pass the _clone_ into `with_default`, and then drop the original `Dispatch` here. Like this:
```rust
        let dispatch = dispatcher::Dispatch::new(subscriber);

        dispatcher::with_default(dispatch.clone(), || {
            // ...
        });

        state.assert_removed("span1");
        state.assert_removed("span2");

        // Ensure the registry itself outlives the span.
        drop(dispatch);
```